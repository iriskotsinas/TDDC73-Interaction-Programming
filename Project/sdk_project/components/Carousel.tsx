import React from 'react';
import {
  View,
  ScrollView,
  Text,
  StyleSheet,
  TouchableOpacity,
  Image,
  Button,
} from 'react-native';
import Modal from 'react-native-modal';

interface ElementProp {
  title: string;
  year: string;
  poster: any;
  rating: string;
  posterurl: string;
}

interface CarouselProps {
  items: Array<ElementProp>;
  itemsPerSlide: number;
}

export const Carousel = (props: CarouselProps) => {
  const {items} = props;
  const itemsPerSlide =
    props.itemsPerSlide === undefined ? 1 : props.itemsPerSlide;

  const [interval, setInterval] = React.useState<number | undefined>(1);
  const [totalIntervals, setTotalIntervals] = React.useState<number>(1);
  const [width, setWidth] = React.useState(0);
  const [modalVisible, setModalVisible] = React.useState<boolean | undefined>(
    false,
  );

  const init = (initialWidth: number) => {
    setWidth(initialWidth);
    if (items === undefined) {
      return;
    }
    const totalItems = items.length;
    setTotalIntervals(Math.ceil(totalItems / itemsPerSlide));
  };

  const getInterval = (rest: number) => {
    for (let i = 1; i <= totalIntervals; i++) {
      if (rest + 1 < (width / totalIntervals) * i) {
        return i;
      }
      if (i === totalIntervals) {
        return i;
      }
    }
  };

  let bullets = [];
  for (let i = 1; i <= totalIntervals; i++) {
    bullets.push(
      <Text
        key={i}
        // eslint-disable-next-line react-native/no-inline-styles
        style={{
          ...styles.bullet,
          opacity: interval === i ? 0.5 : 0.1,
        }}>
        &bull;
      </Text>,
    );
  }

  return (
    <View style={styles.container}>
      <Text>YOUR FILMS</Text>
      <ScrollView
        horizontal={true}
        contentContainerStyle={{
          ...styles.scrollView,
          width: `${100 * totalIntervals}%`,
        }}
        showsHorizontalScrollIndicator={false}
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onContentSizeChange={(w, h) => init(w)}
        onScroll={(data) => {
          setWidth(data.nativeEvent.contentSize.width);
          const thisInterval = getInterval(data.nativeEvent.contentOffset.x);
          setInterval(thisInterval);
        }}
        scrollEventThrottle={200}
        pagingEnabled
        decelerationRate="fast">
        {items &&
          items.map((item: any, index: number) => {
            return (
              <ScrollView key={index}>
                <View>
                  <Modal
                    isVisible={modalVisible}
                    onBackdropPress={() => setModalVisible(true)}>
                    <View style={styles.modal}>
                      <Text
                        style={{
                          ...styles.modalText,
                        }}>{`IMDB-rating: ${item.rating}`}</Text>
                      <Button
                        color="#704646"
                        title="Hide"
                        onPress={() => {
                          setModalVisible(false);
                        }}
                      />
                    </View>
                  </Modal>
                </View>
                <TouchableOpacity
                  testID="totalFilm"
                  key={index}
                  style={styles.stat}
                  onPress={() => {
                    setModalVisible(true);
                  }}>
                  <Image style={styles.image} source={{uri: item.posterurl}} />
                  <Text testID="film" style={{...styles.statText}}>
                    {item.title}
                  </Text>
                  <View style={styles.statHold}>
                    <Text style={{...styles.statLabel}}>{item.year}</Text>
                  </View>
                </TouchableOpacity>
              </ScrollView>
            );
          })}
      </ScrollView>
      <View style={styles.bullets}>{bullets}</View>
    </View>
  );
};

const styles = StyleSheet.create({
  modalText: {
    fontSize: 25,
    marginVertical: 250,
    marginHorizontal: 40,
    textAlign: 'center',
  },
  modal: {
    flex: 1,
    backgroundColor: 'white',
  },
  statsHead: {
    paddingTop: 10,
    paddingHorizontal: 12,
  },
  image: {
    aspectRatio: 2 / 3,
    width: '100%',
  },
  container: {
    width: '100%',
    backgroundColor: '#fbfbfb',
    borderColor: '#ebebeb',
    borderWidth: 1,
    borderRadius: 8,
    shadowColor: '#fcfcfc',
    shadowOpacity: 1,
    marginTop: 10,
    shadowOffset: {
      width: 0,
      height: 5,
    },
  },
  scrollView: {
    display: 'flex',
    flexDirection: 'row',
    overflow: 'hidden',
  },
  bullets: {
    position: 'absolute',
    top: 0,
    right: 0,
    display: 'flex',
    justifyContent: 'flex-start',
    flexDirection: 'row',
    paddingHorizontal: 10,
    paddingTop: 5,
  },
  bullet: {
    paddingHorizontal: 5,
    fontSize: 20,
  },
  stat: {
    paddingHorizontal: 20,
    paddingBottom: 10,
    paddingTop: 30,
    flex: 1,
  },
  statText: {
    width: '100%',
    textAlign: 'left',
    fontSize: 20,
    paddingTop: 20,
  },
  statHold: {
    width: '100%',
    marginBottom: 8,
  },
  statLabel: {
    width: '100%',
    textAlign: 'left',
    fontSize: 11,
    fontWeight: '600',
    paddingTop: 5,
  },
});

export default Carousel;
