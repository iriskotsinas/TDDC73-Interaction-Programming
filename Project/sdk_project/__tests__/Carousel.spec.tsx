import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import {Text, TouchableOpacity, Image} from 'react-native';
import {waitFor} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import Carousel from '../components/Carousel';

configure({adapter: new Adapter()});

const onPressMock = jest.fn();

const items = [
  {
    title: 'Fight Club',
    year: '1999',
    poster:
      'MV5BZGY5Y2RjMmItNDg5Yy00NjUwLThjMTEtNDc2OGUzNTBiYmM1XkEyXkFqcGdeQXVyNjU0OTQ0OTY@._V1_SX290_CR0,0,290,429_AL_.jpg',
    rating: '8.8',
    posterurl:
      'https://m.media-amazon.com/images/M/MV5BMmEzNTkxYjQtZTc0MC00YTVjLTg5ZTEtZWMwOWVlYzY0NWIwXkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_UX182_CR0,0,182,268_AL_.jpg',
  },
  {
    title: 'The Dark Knight',
    year: '2008',
    poster:
      'MV5BMTMxNTMwODM0NF5BMl5BanBnXkFtZTcwODAyMTk2Mw@@._V1_SY500_CR0,0,337,500_AL_.jpg',
    rating: '9.0',
    posterurl:
      'https://images-na.ssl-images-amazon.com/images/M/MV5BMTMxNTMwODM0NF5BMl5BanBnXkFtZTcwODAyMTk2Mw@@._V1_SY500_CR0,0,337,500_AL_.jpg',
  },
];

describe('<Carousel />', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Carousel onPress={() => {}} itemsPerSlide={2} items={items} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with default settings', (): void => {
    const wrapper = shallow(
      <Carousel onPress={() => {}} itemsPerSlide={2} items={items} />,
    );
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render all items in items', async () => {
    const wrapper = shallow(
      <Carousel onPress={() => {}} itemsPerSlide={2} items={items} />,
    );

    // Two items rendered
    const carousel = wrapper.find({testID: 'film'});
    expect(carousel.length).toBe(items.length);
  });

  it('should render correct titles', () => {
    const wrapper = shallow(
      <Carousel onPress={() => {}} itemsPerSlide={2} items={items} />,
    );

    wrapper.find({testID: 'film'}).forEach((node, index) => {
      expect(node.props().children).toBe(items[index].title);
    });
  });

  it('shows rating after film poster is pressed', async () => {
    let inst = renderer.create(
      <Carousel onPress={() => {}} itemsPerSlide={2} items={items} />,
    );

    await waitFor(() => {
      const button = inst.root.findAllByType(TouchableOpacity)[0];
      const text = inst.root.findAllByType(Text)[1];

      button.props.onPress();

      expect(text.props.children).toBe('IMDB-rating: ' + items[0].rating);
      // expect(button.props.children)
    });
  });

  it('shows same amount of film posters as items in the list', async () => {
    onPressMock.mockReturnValue(false);
    let inst = renderer.create(
      <Carousel onPress={() => {}} itemsPerSlide={2} items={items} />,
    );

    const sizeItems = inst.root.findAllByType(Image);
    expect(sizeItems.length).toEqual(items.length);
  });
  
});
