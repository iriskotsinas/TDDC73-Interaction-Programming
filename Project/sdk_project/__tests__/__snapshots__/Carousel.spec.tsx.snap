// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fbfbfb",
      "borderColor": "#ebebeb",
      "borderRadius": 8,
      "borderWidth": 1,
      "marginTop": 10,
      "shadowColor": "#fcfcfc",
      "shadowOffset": Object {
        "height": 5,
        "width": 0,
      },
      "shadowOpacity": 1,
      "width": "100%",
    }
  }
>
  <Text>
    YOUR FILMS
  </Text>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "overflow": "hidden",
        "width": "100%",
      }
    }
    decelerationRate="fast"
    horizontal={true}
    onContentSizeChange={[Function]}
    onScroll={[Function]}
    pagingEnabled={true}
    scrollEventThrottle={200}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <RCTScrollView>
        <View>
          <View>
            <Modal
              animationType="none"
              hardwareAccelerated={false}
              hideModalContentWhileAnimating={false}
              onModalHide={[Function]}
              onModalWillHide={[Function]}
              onModalWillShow={[Function]}
              onRequestClose={[Function]}
              scrollHorizontal={false}
              scrollOffset={0}
              scrollOffsetMax={0}
              scrollTo={null}
              supportedOrientations={
                Array [
                  "portrait",
                  "landscape",
                ]
              }
              swipeThreshold={100}
              transparent={true}
              visible={false}
            >
              <View
                accessible={true}
                focusable={true}
                forwardedRef={[Function]}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "black",
                    "bottom": 0,
                    "height": 1334,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "width": 750,
                  }
                }
              />
              <View
                forwardedRef={[Function]}
                hideModalContentWhileAnimating={false}
                onModalHide={[Function]}
                onModalWillHide={[Function]}
                onModalWillShow={[Function]}
                pointerEvents="box-none"
                scrollHorizontal={false}
                scrollOffset={0}
                scrollOffsetMax={0}
                scrollTo={null}
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                    "margin": 37.5,
                    "transform": Array [
                      Object {
                        "translateY": 0,
                      },
                    ],
                  }
                }
                supportedOrientations={
                  Array [
                    "portrait",
                    "landscape",
                  ]
                }
                swipeThreshold={100}
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "white",
                      "flex": 1,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "fontSize": 25,
                        "marginHorizontal": 40,
                        "marginVertical": 250,
                        "textAlign": "center",
                      }
                    }
                  >
                    IMDB-rating: 8.8
                  </Text>
                  <View
                    accessibilityRole="button"
                    accessibilityState={Object {}}
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {},
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#007AFF",
                              "fontSize": 18,
                              "margin": 8,
                              "textAlign": "center",
                            },
                            Object {
                              "color": "#704646",
                            },
                          ]
                        }
                      >
                        Hide
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </Modal>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "opacity": 1,
                "paddingBottom": 10,
                "paddingHorizontal": 20,
                "paddingTop": 30,
              }
            }
            testID="totalFilm"
          >
            <Image
              source={
                Object {
                  "uri": "https://m.media-amazon.com/images/M/MV5BMmEzNTkxYjQtZTc0MC00YTVjLTg5ZTEtZWMwOWVlYzY0NWIwXkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_UX182_CR0,0,182,268_AL_.jpg",
                }
              }
              style={
                Object {
                  "aspectRatio": 0.6666666666666666,
                  "width": "100%",
                }
              }
            />
            <Text
              style={
                Object {
                  "fontSize": 20,
                  "paddingTop": 20,
                  "textAlign": "left",
                  "width": "100%",
                }
              }
              testID="film"
            >
              Fight Club
            </Text>
            <View
              style={
                Object {
                  "marginBottom": 8,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 11,
                    "fontWeight": "600",
                    "paddingTop": 5,
                    "textAlign": "left",
                    "width": "100%",
                  }
                }
              >
                1999
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <RCTScrollView>
        <View>
          <View>
            <Modal
              animationType="none"
              hardwareAccelerated={false}
              hideModalContentWhileAnimating={false}
              onModalHide={[Function]}
              onModalWillHide={[Function]}
              onModalWillShow={[Function]}
              onRequestClose={[Function]}
              scrollHorizontal={false}
              scrollOffset={0}
              scrollOffsetMax={0}
              scrollTo={null}
              supportedOrientations={
                Array [
                  "portrait",
                  "landscape",
                ]
              }
              swipeThreshold={100}
              transparent={true}
              visible={false}
            >
              <View
                accessible={true}
                focusable={true}
                forwardedRef={[Function]}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "black",
                    "bottom": 0,
                    "height": 1334,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "width": 750,
                  }
                }
              />
              <View
                forwardedRef={[Function]}
                hideModalContentWhileAnimating={false}
                onModalHide={[Function]}
                onModalWillHide={[Function]}
                onModalWillShow={[Function]}
                pointerEvents="box-none"
                scrollHorizontal={false}
                scrollOffset={0}
                scrollOffsetMax={0}
                scrollTo={null}
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                    "margin": 37.5,
                    "transform": Array [
                      Object {
                        "translateY": 0,
                      },
                    ],
                  }
                }
                supportedOrientations={
                  Array [
                    "portrait",
                    "landscape",
                  ]
                }
                swipeThreshold={100}
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "white",
                      "flex": 1,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "fontSize": 25,
                        "marginHorizontal": 40,
                        "marginVertical": 250,
                        "textAlign": "center",
                      }
                    }
                  >
                    IMDB-rating: 9.0
                  </Text>
                  <View
                    accessibilityRole="button"
                    accessibilityState={Object {}}
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {},
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#007AFF",
                              "fontSize": 18,
                              "margin": 8,
                              "textAlign": "center",
                            },
                            Object {
                              "color": "#704646",
                            },
                          ]
                        }
                      >
                        Hide
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </Modal>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "opacity": 1,
                "paddingBottom": 10,
                "paddingHorizontal": 20,
                "paddingTop": 30,
              }
            }
            testID="totalFilm"
          >
            <Image
              source={
                Object {
                  "uri": "https://images-na.ssl-images-amazon.com/images/M/MV5BMTMxNTMwODM0NF5BMl5BanBnXkFtZTcwODAyMTk2Mw@@._V1_SY500_CR0,0,337,500_AL_.jpg",
                }
              }
              style={
                Object {
                  "aspectRatio": 0.6666666666666666,
                  "width": "100%",
                }
              }
            />
            <Text
              style={
                Object {
                  "fontSize": 20,
                  "paddingTop": 20,
                  "textAlign": "left",
                  "width": "100%",
                }
              }
              testID="film"
            >
              The Dark Knight
            </Text>
            <View
              style={
                Object {
                  "marginBottom": 8,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 11,
                    "fontWeight": "600",
                    "paddingTop": 5,
                    "textAlign": "left",
                    "width": "100%",
                  }
                }
              >
                2008
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingHorizontal": 10,
        "paddingTop": 5,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "opacity": 0.5,
          "paddingHorizontal": 5,
        }
      }
    >
      •
    </Text>
  </View>
</View>
`;

exports[`<Carousel /> should render with default settings 1`] = `ShallowWrapper {}`;
